File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/AggregateMeta.java
package io.dddrive.core.ddd.model;

import java.time.OffsetDateTime;
import java.util.List;

import io.dddrive.core.oe.model.ObjUser;
import io.dddrive.core.property.model.PropertyChangeListener;
import io.dddrive.core.validation.model.AggregatePartValidation;

/**
 * A DDD Aggregate Root Meta Information.
 */
public interface AggregateMeta {

	AggregateRepository<?> getRepository();

	boolean isFrozen();

	boolean isInCalc();

	boolean isCalcEnabled();

	void disableCalc();

	void enableCalc();

	boolean isInLoad();

	void beginLoad();

	void endLoad();

	void addPropertyChangeListener(PropertyChangeListener listener);

	void removePropertyChangeListener(PropertyChangeListener listener);

	Integer getVersion();

	List<AggregatePartValidation> getValidations();

	OffsetDateTime getCreatedAt();

	ObjUser getCreatedByUser();

	OffsetDateTime getModifiedAt();

	ObjUser getModifiedByUser();

}

File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/AggregateMeta.kt
package io.dddrive.core.ddd.model

import io.dddrive.core.oe.model.ObjUser
import io.dddrive.core.property.model.PropertyChangeListener
import io.dddrive.core.validation.model.AggregatePartValidation
import java.time.OffsetDateTime

/**
 * A DDD Aggregate Root Meta Information.
 */
interface AggregateMeta : EntityMeta {

	val repository: AggregateRepository<*>

	fun beginLoad()

	fun endLoad()

	fun addPropertyChangeListener(listener: PropertyChangeListener)

	fun removePropertyChangeListener(listener: PropertyChangeListener)

	val version: Int

	val validations: List<AggregatePartValidation>

	val createdAt: OffsetDateTime

	val createdByUser: ObjUser

	val modifiedAt: OffsetDateTime?

	val modifiedByUser: ObjUser?

}
